### PR TITLE
AirfRANS 4L/320d: intermediate width scaling with golden config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The 4L/256d architecture achieved a 22.3% improvement over 3L/192d, suggesting that width scaling is a productive direction. 4L/320d is the natural next step — an intermediate width between 256d and 384d (which would have 6 heads) that keeps compute tractable. With 5 heads at 320d (64d per head), attention remains efficient and the model gets ~56% more hidden-state capacity than 256d. The golden regularization config (WD=1e-2, gc=1.0, T_max=5) should generalize.

## Instructions
Run the following command exactly:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 \
  --wandb_group airfrans-arch-scaling
```

Key changes from golden config: `--model-hidden-dim 320 --model-heads 5` (was 256d/4 heads). All other settings match the proven 4L/256d golden config. Note: 320/5 = 64 dimensions per head, same ratio as 256/4.

## Baseline
Current best metrics from BASELINE.md (PR #2727, W&B run: ruurxdqs):
- val_primary/surface_mse: **0.007264** (primary metric, lower is better)
- External target: 0.0043 (~1.7x gap remaining)
- Architecture: 4L/256d, 4 heads, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5, no-EMA, enable-fourier

Reproduce baseline:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chihiro:

## Results

### 4L/320d + golden config (lr=7e-4, gc=1.0, WD=1e-2, T_max=5)

| Metric | Baseline (4L/256d) | 4L/320d |
|--------|-------------------|---------|
| **val_primary/surface_mse** | 0.007264 | **0.0720** (best, ep 15) |
| test_primary/surface_mse | — | 0.2227 (final ep) |
| Epochs | many | 28 (diverged) |

**Result: NEGATIVE.** The model diverged. Best val is 9.9x worse than baseline.

### Epoch progression — gradient explosion

| Epoch | val_mse | loss | grad_norm_mean |
|-------|---------|------|----------------|
| 1 | 0.4202 | 1.444 | 12.0 |
| 5 | 0.2596 | 0.767 | 17.4 |
| 10 | 0.1398 | 0.584 | 16.7 |
| **15** | **0.0720** (best) | 0.456 | 39.2 |
| 20 | 0.2182 | 0.616 | 61.7 |
| 22 | 0.2525 | 0.881 | **467.5** |
| 25 | 0.2686 | 0.962 | **2113.9** |
| 28 | 0.2452 | 1.099 | **2964.8** |

### Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 --wandb-group airfrans-arch-scaling \
  --agent chihiro --wandb-name "chihiro/airfrans-4L320d-golden"
```

- **W&B run:** pmwf5itv
- **Peak memory:** ~7 GB VRAM

### What happened

**Training diverged due to gradient explosion.** Gradient norms grew steadily from ~12 at epoch 1 to ~40 at epoch 15, then exploded to 2965 by epoch 28. The grad_clip at 1.0 was unable to contain the instability — nearly every batch triggered clipping (360/360 steps) even at epoch 1.

The best val (0.072 at epoch 15) is **9.9x worse** than the 4L/256d baseline (0.007264). Even the best epoch showed high noise — val bounced from 0.072 to 0.157 between epochs 15-16.

**Root cause: lr=7e-4 is too aggressive for 4L/320d.** The wider model has ~56% more parameters than 4L/256d, which amplifies gradient magnitudes. The golden config's lr=7e-4 appears to be at or near the stability boundary for 4L/256d, and scaling width pushes past it. T_max=5 compounds the problem — the LR stays near peak for most of the cycle, giving gradients no low-LR recovery window.

### Suggested follow-ups

1. **Lower LR:** Try lr=5e-4 or lr=3e-4 with 4L/320d — scale LR inversely with model width.
2. **Stronger grad clip:** gc=0.5 or gc=0.3 to contain early gradient growth.
3. **Larger T_max:** T_max=10 or T_max=15 gives more time at low LR to stabilize between restarts.
4. **Combined fix:** lr=5e-4 + gc=0.5 + T_max=10 — attack all three instability factors.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2780 because ADVISOR: Closing — catastrophic divergence at 4L/320d with golden config lr=7e-4. Grad norms 12→2965 by epoch 28, best val 0.0720 (9.9x worse than baseline 0.007264). Root cause: lr=7e-4 is near the stability boundary for 4L/256d, and scaling to 4L/320d (+56% parameters) amplifies gradient instability past the threshold. T_max=5 compounds it (no low-LR recovery window). If 4L/320d is attempted again, use lr≤5e-4 with gc≤0.5. NOTE: The pressure-weight breakthrough (PR #2703, 0.00435) has made architecture scaling experiments less critical — loss reformulation is the dominant lever now..
